backend/layers/processing/rollback.py

```diff
@@ -0,0 +1,301 @@
+            rolled_back_collection_versions.append(rolled_back_collection_version)
+        self._clean_up_published_collection_versions(rolled_back_collection_versions)
+
+    def rollback_public_collection(self, collection_id: CollectionId) -> CollectionVersion:
+        """
+        For a given public CollectionId, rolls back the associated Collection to its previous, most recently published
+        CollectionVersion (if one exists). This will restore the state of the public collection to its pre-migration
+        state.
+
+        :param collection_id: CollectionId of the Collection to rollback
+        :return: CollectionVersion that was rolled back
+        """
+        return self.business_logic.restore_previous_collection_version(collection_id)
+
+    def _clean_up_rolled_back_datasets(self, rolled_back_datasets: List[DatasetVersion]) -> None:
+        """
+        Triggers deletion of the DB references and S3 assets for the rolled back DatasetVersions.
+        """
+        self.business_logic.delete_dataset_versions(rolled_back_datasets)
+
+    def _clean_up_published_collection_versions(self, rolled_back_collection_versions: List[CollectionVersion]) -> None:
+        """
+        Triggers deletion of the DB references and S3 assets for the rolled back CollectionVersions' associated
+        DatasetVersions, if they are not associated with any still-existing CollectionVersions.
+        """
+        datasets_to_rollback = []
+        for rolled_back_collection_version in rolled_back_collection_versions:
+            dataset_rollback_candidates = rolled_back_collection_version.datasets
+            collection_version_history = self.business_logic.get_collection_versions_from_canonical(
+                rolled_back_collection_version.collection_id
+            )
+            dataset_version_history = {dv.version_id.id: dv for cv in collection_version_history for dv in cv.datasets}
+            for dataset_rollback_candidate in dataset_rollback_candidates:
+                if dataset_rollback_candidate.version_id.id not in dataset_version_history:
+                    datasets_to_rollback.append(dataset_rollback_candidate)
+        self._clean_up_rolled_back_datasets(datasets_to_rollback)
+
+
+if __name__ == "__main__":
+    business_logic = BusinessLogic(
+        DatabaseProvider(),
+        None,
+        None,
+        None,
+        S3Provider(),
+        UriProvider(),
+    )
+    rollback_type_str = os.environ.get("ROLLBACK_TYPE", None)
+    if rollback_type_str is None:
+        raise ValueError("ROLLBACK_TYPE is required")
+
+    rollback_type = RollbackType(rollback_type_str)
+    if rollback_type in (
+        RollbackType.PUBLIC_COLLECTION_LIST,
+        RollbackType.PRIVATE_COLLECTION_LIST,
+        RollbackType.PRIVATE_DATASET_LIST,
+    ):
+        entities_to_rollback = os.environ.get("ENTITY_LIST", None)
+        entities_to_rollback_list = entities_to_rollback.split(",") if entities_to_rollback else None
```

Comment:
Is there a use case to provide no Entities? If ti is a no op, I think it would be a better signal to throw an error. That should make it clear to the user that they forgot to add entities.